kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncThreadPool.java

```diff
@@ -135,21 +142,29 @@ public <KOut, VOut> void scheduleForProcessing(
     final var inFlightKey = InFlightWorkKey.of(processorName, partition);
     final var inFlightForTask
         = inFlight.computeIfAbsent(inFlightKey, k -> new ConcurrentHashMap<>());
-    for (final var e : events) {
+
+    for (final AsyncEvent event : events) {
       final var future = CompletableFuture.runAsync(
-          new AsyncEventTask<>(e, taskContext, asyncProcessorContext, finalizingQueue),
+          new AsyncEventTask<>(event, taskContext, asyncProcessorContext, finalizingQueue),
           executor
       );
       final var inFlightEvent = new InFlightEvent(future);
-      inFlightForTask.put(e, inFlightEvent);
-      future.handle((r, t) -> {
-        if (t != null) {
-          inFlightEvent.setError(t);
-        } else {
-          inFlightForTask.remove(e);
-        }
-        return null;
-      });
+      inFlightForTask.put(event, inFlightEvent);
+
+      future
+          .whenComplete((r, t) -> inFlightForTask.remove(event))
+          .exceptionally(processingException -> {
+            inFlightEvent.setError(processingException);
+            event.transitionToFailed(processingException);
+            finalizingQueue.addFailedEvent(event, processingException);
+            return null;
+          })
+          .exceptionally(processingException -> {
+            // do this in separate stage to ensure we always catch a fatal exception, even
+            // if we somehow hit another exception in the previous exception handling stage
+            fatalExceptions.computeIfAbsent(inFlightKey, k -> processingException);
```

Comment:
This is our "failsafe" for any exceptions that don't end up in the finalizing queue. In theory that should never happen, but it's technically possible for something in the previous `exceptionally` block to throw